totalRP3/modules/NamePlates/NamePlates.lua

```diff
@@ -0,0 +1,1124 @@
+		local ownerID = TRP3_Utils.str.companionIDToInfo(registerUnitID);
+		if TRP3_Register.isUnitIDKnown(ownerID) then
+			TRP3_API.r.sendQuery(ownerID);
+		end
+	end
+
+	-- Debounce further requests for a significant amount of time.
+	local cooldown = GetTime() + ACTIVE_QUERY_COOLDOWN;
+	self.registerUnitIDCooldowns[registerUnitID] = cooldown;
+end
+
+-- Returns the name text to be displayed for the given unit token, or nil
+-- if customizations are disabled or no profile is available.
+function TRP3_NamePlates:GetCustomUnitName(unitToken)
+	-- If customization is disabled entirely for any reason, stop.
+	if not self:ShouldCustomizeUnitFrames() then
+		return nil;
+	end
+
+	-- Dispatch based on the profile type.
+	if UnitIsPlayer(unitToken) and self:ShouldShowCustomPlayerNames() then
+		-- Get the profile for the player and with it, their name.
+		local profile = GetPlayerProfile(unitToken);
+		if not profile then
+			return nil;
+		end
+
+		local nameText = profile:GetRoleplayingName();
+
+		-- Prefix the OOC indicator if configured.
+		if not profile:IsInCharacter() and self:ShouldShowOOCIndicators() then
+			local oocIndicator = self:GetConfiguredOOCIndicator();
+			if oocIndicator == "TEXT" then
+				nameText = strjoin(" ", OOC_TEXT_INDICATOR, nameText);
+			elseif oocIndicator == "ICON" then
+				nameText = strjoin(" ", tostring(OOC_ICON_INDICATOR), nameText);
+			end
+		end
+
+		return nameText;
+	elseif UnitIsCombatPet(unitToken) and self:ShouldShowCustomPetNames() then
+		-- Combat pets use companion pet profiles.
+		local profile = GetCombatPetProfile(unitToken);
+		if not profile then
+			return nil;
+		end
+
+		return profile.NA;
+	end
+
+	-- Unknown profile type.
+	return nil;
+end
+
+-- Returns the custom color to display for the given unit token, or nil if
+-- customization is disabled or no profile is available.
+function TRP3_NamePlates:GetCustomUnitColor(unitToken)
+	-- If customization is disabled entirely for any reason, stop.
+	if not self:ShouldCustomizeUnitFrames() then
+		return nil;
+	end
+
+	-- Don't bother if custom colors are disabled.
+	if not self:ShouldShowCustomColors() then
+		return nil;
+	end
+
+	-- Dispatch based on the profile type.
+	if UnitIsPlayer(unitToken) then
+		-- Get the profile for the player and with it, their custom color.
+		local profile = GetPlayerProfile(unitToken);
+		if not profile then
+			return nil;
+		end
+
+		local nameColor = profile:GetCustomColorForDisplay();
+
+		-- If there is no color, use class coloring instead.
+		if not nameColor then
+			local _, class = UnitClass(unitToken);
+			if class then
+				nameColor = C_ClassColor.GetClassColor(class);
+			end
+		end
+
+		return nameColor;
+	elseif UnitIsCombatPet(unitToken) then
+		-- Combat pets use companion pet profiles.
+		local profile = GetCombatPetProfile(unitToken);
+		if not profile then
+			return nil;
+		end
+
+		return profile.NH and Color.CreateFromHexa(profile.NH) or nil;
```

Comment:
This will not provide a color suitable for display.
```suggestion
		local petColor = profile.NH and Color.CreateFromHexa(profile.NH);
		if petColor and AddOn_TotalRP3.Configuration.shouldDisplayIncreasedColorContrast() then
			petColor:LightenColorUntilItIsReadableOnDarkBackgrounds();
		end
		return petColor;
```